Core/TransactionHandler.php

```diff
@@ -0,0 +1,314 @@
+ */
+
+namespace Wirecard\Oxid\Core;
+
+use OxidEsales\Eshop\Application\Model\Order;
+use OxidEsales\Eshop\Core\DatabaseProvider;
+use OxidEsales\Eshop\Core\Registry;
+use OxidEsales\Eshop\Core\Field;
+
+use Wirecard\Oxid\Model\PaymentMethod;
+use Wirecard\Oxid\Model\Transaction;
+use Wirecard\PaymentSdk\Entity\Amount;
+use Wirecard\PaymentSdk\Response\SuccessResponse;
+use Wirecard\PaymentSdk\Response\FailureResponse;
+use Wirecard\PaymentSdk\Response\Response;
+use Wirecard\PaymentSdk\BackendService;
+
+use Exception;
+
+/**
+ * Class TransactionHandler
+ * @mixin Wirecard\Oxid\Core
+ *
+ * @since 1.0.1
+ */
+class TransactionHandler
+{
+    /**
+     * @var \Psr\Log\LoggerInterface
+     *
+     * @since 1.0.1
+     */
+    private $_oLogger;
+
+    /**
+     * TransactionHandler constructor.
+     *
+     * @since 1.0.1
+     */
+    public function __construct()
+    {
+        $this->_oLogger = Registry::getLogger();
+    }
+
+    /**
+     * Executes a post-processing action on a transaction.
+     *
+     * @param Transactio $oParentTransaction
+     * @param string     $sActionTitle
+     * @param float      $fAmount
+     *
+     * @return array either [status => success] or [status => error, message => errorMessage]
+     *
+     * @since 1.0.1
+     */
+    public function processAction($oParentTransaction, $sActionTitle, $fAmount = null)
+    {
+        $oPaymentMethod = $this->_getPaymentMethod($oParentTransaction);
+        $oOrder = $oParentTransaction->getTransactionOrder();
+        $oConfig = $oPaymentMethod->getConfig($oOrder->getOrderPayment());
+
+        $oTransaction = $oPaymentMethod->getTransaction();
+
+        $sParentTransactionId = $oParentTransaction->wdoxidee_ordertransactions__transactionid->value;
+        $oTransaction->setParentTransactionId($sParentTransactionId);
+
+        if (!is_null($fAmount)) {
+            $sCurrencyName = $oOrder->oxorder__oxcurrency->value;
+            $oCurrency = Registry::getConfig()->getCurrencyObject($sCurrencyName);
+
+            $oTransaction->setAmount(new Amount(
+                Registry::getUtils()->fRound($fAmount, $oCurrency),
+                $sCurrencyName
+            ));
+        }
+
+        // use the backend service to process the action
+        $oBackendService = new BackendService($oConfig, $this->_oLogger);
+        $oResponse = $oBackendService->process($oTransaction, $sActionTitle);
+
+        return $this->_handleProcessActionResponse($oResponse, $oOrder, $oBackendService);
+    }
+
+    /**
+     * Handles the success and error response from the post-processing action.
+     *
+     * @param Response       $oResponse
+     * @param Order          $oOrder
+     * @param BackendService $oBackendService
+     *
+     * @return array either [status => success] or [status => error, message => errorMessage]
+     *
+     * @since 1.0.1
+     */
+    private function _handleProcessActionResponse($oResponse, $oOrder, $oBackendService)
```

Comment:
It should be enough to just call it `_onActionResponse` instead.

Also, it would be cleaner to split this function into a `_onActionFailure` and `_onActionSuccess` function.

Reply:
Very good point, refactored it 👍